google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudApiManager.java

```diff
@@ -119,6 +146,175 @@ static void enableApis(Set<CloudLibrary> libraries, CloudProject cloudProject, P
+            progress,
+            GctBundle.message("cloud.apis.service.account.add.roles.progress.message"),
+            step / numSteps);
+        step++;
+        addRolesToServiceAccount(user.get(), serviceAccount, roles, cloudProject);
+      }
+
+      showProgress(
+          progress,
+          GctBundle.message("cloud.apis.service.account.create.key.progress.message"),
+          step / numSteps);
+      step++;
+      ServiceAccountKey serviceAccountKey = createServiceAccountKey(user.get(), serviceAccount);
+
+      showProgress(
+          progress,
+          GctBundle.message("cloud.apis.service.account.download.key.progress.message"),
+          step / numSteps);
+      Path keyPath = writeServiceAccountKey(serviceAccountKey, downloadDir, cloudProject);
+
+      notifyServiceAccountCreated(project, name, keyPath);
+    } catch (IOException e) {
+      LOG.warn(
+          "Exception occurred attempting to create service account on GCP and download its key", e);
+    }
+  }
+
+  /** Creates a new {@link ServiceAccount} for the given {@link CloudProject} using the IAM API. */
+  private static ServiceAccount createServiceAccount(
+      CredentialedUser user, String name, CloudProject cloudProject) throws IOException {
+    CreateServiceAccountRequest request = new CreateServiceAccountRequest();
+    ServiceAccount serviceAccount = new ServiceAccount();
+    serviceAccount.setDisplayName(name);
+    request.setServiceAccount(serviceAccount);
+    request.setAccountId(createServiceAccountId(name));
+
+    Iam iam = GoogleApiClientFactory.getInstance().getIamClient(user.getCredential());
+
+    return iam.projects()
+        .serviceAccounts()
+        .create(
+            String.format(SERVICE_ACCOUNT_CREATE_REQUEST_PROJECT_PATTERN, cloudProject.projectId()),
+            request)
+        .execute();
+  }
+
+  /**
+   * Adds a set of {@link Role roles} to a {@link ServiceAccount}.
+   *
+   * <p>This is done by fetching the cloud project's existing IAM Policy, adding the new roles to
+   * the given service account, and then writing the updated policy back to the cloud project.
+   *
+   * @param user the current {@link CredentialedUser}
+   * @param serviceAccount the {@link ServiceAccount} to which to add roles
+   * @param roles the set of {@link Role} to be added to the service account
+   * @param cloudProject the current {@link CloudProject}
+   * @throws IOException if the API call fails to update the IAM policy
+   */
+  private static void addRolesToServiceAccount(
+      CredentialedUser user,
+      ServiceAccount serviceAccount,
+      Set<Role> roles,
+      CloudProject cloudProject)
+      throws IOException {
+    CloudResourceManager resourceManager =
+        GoogleApiClientFactory.getInstance().getCloudResourceManagerClient(user.getCredential());
+
+    Policy existingPolicy =
+        resourceManager
+            .projects()
+            .getIamPolicy(cloudProject.projectId(), new GetIamPolicyRequest())
```

Comment:
Looks like this is a policy for a whole cloud project, I guess there is no way to get a policy and bindings only for a specific service account, correct?

Reply:
my understanding is that the Policy is at the cloud project level. The bindings (where the roles are defined), however, can be scoped to a service account. This is why in the next few lines of code, the new bindings with the new roles are scoped to the service account (meaning it is the service account that will be requesting access to the resources).

I will double check this though.